admin/worker/run_autoscaler.go

```diff
@@ -18,10 +30,39 @@ func (w *Worker) runAutoscaler(ctx context.Context) error {
 	}
 
 	for _, rec := range recs {
-		// TODO: Add autoscaling logic based on the recommendation here.
-		// Consider checking rec.UpdatedOn to avoid making autoscaling decision on stale data!
-		w.logger.Info("autoscaler recommendation", zap.String("project_id", rec.ProjectID), zap.Int("recommended_slots", rec.RecommendedSlots))
-		break
+		// if UpdatedOn is too old (7 days), the recommendation is stale and may not be trusted.
+		if time.Since(rec.UpdatedOn) >= legacyRecommendTime {
+			w.logger.Debug("skipping autoscaler: the recommendation is stale", zap.String("project_id", rec.ProjectID), zap.Time("recommendation_updated_on", rec.UpdatedOn))
+			break
+		}
+
+		targetProject, err := w.admin.DB.FindProject(ctx, rec.ProjectID)
+		if err != nil {
+			w.logger.Debug("failed to find project:", zap.String("project_id", rec.ProjectID), zap.Error(err))
+			break
+		}
+
+		if !shouldScale(targetProject.ProdSlots, rec.RecommendedSlots) {
+			w.logger.Debug("skipping autoscaler: target slots are within threshold of original slots",
+				zap.Int("project_slots", targetProject.ProdSlots),
+				zap.Int("recommend_slots", rec.RecommendedSlots),
+				zap.Float64("scale_threshold_percentage", scaleThreshold),
+				zap.String("project_id", targetProject.ID),
+			)
+			break
+		}
+
+		opts := &database.UpdateProjectOptions{
```

Comment:
Can we add an extra check here for `if rec.RecommendedSlots <= 0` just to be completely safe? (The recommender should always return `>= 1`, but just in case something went wrong there.)

Reply:
Sure thing! Just added.